src/Configuration.php

```diff
@@ -0,0 +1,124 @@
+
+// @TODO should the naming be swapped around? `new BetterReflection()` sounds MUCH cooler!
+final class Configuration implements BetterReflection
+{
+    /**
+     * @var SourceLocator|null
+     */
+    private $sourceLocator;
+
+    /**
+     * @var ClassReflector|null
+     */
+    private $classReflector;
+
+    /**
+     * @var FunctionReflector|null
+     */
+    private $functionReflector;
+
+    /**
+     * @var Parser|null
+     */
+    private $phpParser;
+
+    /**
+     * @var AstLocator|null
+     */
+    private $astLocator;
+
+    /**
+     * @var AstConversionStrategy|null
+     */
+    private $astConversionStrategy;
+
+    /**
+     * @var CompileNodeToValue|null
+     */
+    private $compileNodeToValue;
+
+    /**
+     * @var FindReflectionOnLine|null
+     */
+    private $findReflectionOnLine;
+
+    public function sourceLocator() : SourceLocator
+    {
+        $astLocator = $this->astLocator();
+        $parser     = $this->phpParser();
+
+        return $this->sourceLocator
+            ?? $this->sourceLocator = new AggregateSourceLocator([
+                new PhpInternalSourceLocator($astLocator, $parser),
+                new EvaledCodeSourceLocator($astLocator, $parser),
+                new AutoloadSourceLocator($astLocator),
+            ]);
+    }
+
+    public function classReflector() : ClassReflector
+    {
+        return $this->classReflector
+            ?? $this->classReflector = new ClassReflector($this->sourceLocator());
+    }
+
+    public function functionReflector() : FunctionReflector
+    {
+        return $this->functionReflector
+            ?? $this->functionReflector = new FunctionReflector($this->sourceLocator(), $this->classReflector());
+    }
+
+    public function phpParser() : Parser
+    {
+        return $this->phpParser
+            ?? $this->phpParser = (new ParserFactory())->create(ParserFactory::PREFER_PHP7, new Emulative([
+                'usedAttributes' => ['comments', 'startLine', 'endLine', 'startFilePos', 'endFilePos'],
+            ]));
+    }
+
+    public function astLocator() : AstLocator
+    {
+        return $this->astLocator
+            ?? $this->astLocator = new AstLocator($this->phpParser());
+    }
+
+    public function astConversionStrategy() : AstConversionStrategy
+    {
+        return $this->astConversionStrategy
+            ?? $this->astConversionStrategy = new NodeToReflection();
+    }
+
+    public function compileNodeToValue() : CompileNodeToValue
+    {
+        return $this->compileNodeToValue
+            ?? $this->compileNodeToValue = new CompileNodeToValue();
+    }
+
+    public function findReflectionsOnLine() : FindReflectionOnLine
+    {
+        return $this->findReflectionOnLine
+            ?? $this->findReflectionOnLine = new FindReflectionOnLine($this->sourceLocator(), $this->astLocator());
+    }
+
+    // @TODO modifier methods here. They should be `withFoo()`, `withBar()`, and be immutable.
```

Comment:
Don't know if we should make this configurable at all, at this point. After implementing the entire setup in the test suite, this degree of configuration was no longer needed.